scylla-cql/src/frame/value.rs

```diff
@@ -51,6 +51,81 @@ pub enum MaybeUnset<V> {
     Set(V),
 }
 
+/// Native CQL `varint` representation.
+///
+/// Represented as two's-complement binary in big-endian order.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct CqlVarint(Vec<u8>);
+
+/// Constructors from bytes
+impl CqlVarint {
+    /// Creates a [`CqlVarint`] from an array of bytes in
+    /// two's complement big-endian binary representation.
+    pub fn from_signed_bytes_be(mut digits: Vec<u8>) -> Self {
+        Self::normalize(&mut digits);
+        Self(digits)
+    }
+
+    /// Creates a [`CqlVarint`] from a slice of bytes in
+    /// two's complement binary big-endian representation.
+    pub fn from_signed_bytes_be_slice(digits: &[u8]) -> Self {
+        Self::from_signed_bytes_be(digits.to_vec())
+    }
+
+    /// Performs a normalization on bytes provided in the constructor.
+    ///
+    /// Removes leading zeros from the array.
+    /// The only exception is when the first non-zero digit is greater
+    /// than 0x7F (msbit is 1). In such case, one leading zero digit is left.
+    fn normalize(digits: &mut Vec<u8>) {
+        digits.reverse();
+
+        let mut has_leading_zero = false;
+        while let Some(0x00) = digits.last() {
+            digits.pop();
+            has_leading_zero = true;
+        }
+
+        match digits.last() {
+            // Positive number. Most-significant bit is 1, so we need to append a 0x00 byte.
+            Some(x) if x > &0x7f && has_leading_zero => digits.push(0x00),
+            // Negative number - leave as is.
+            Some(_) => (),
+            // Empty vector (or filled with zeros) - represent as 0.
+            None => digits.push(0x00),
+        }
+
+        digits.reverse();
```

Comment:
Using `VecDeque` would allow to skip those `reverse()` calls. Otoh it would require `[make_contiguous](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.make_contiguous)` call in `as_signed_bytes_be_slice`, so there would be some overhead on first call. Wdyt?

Reply:
It would also require that `as_signed_bytes_be_slice` accepts `&mut self` instead of `&self`. This looks a bit strange to me.

I'm really not sure which approach is better here. Maybe @piodul has a stronger opinion on that.

Reply:
No need to move data around so much or involve `VecDeque`. I propose this algorithm:

- Find the first non-zero digit `d` and its position `p`,
- If all digits are zero, resize/truncate the vec to size 1,
- If `d > 0x7f` then cut the first `p - 1` chars,
- Otherwise, cut the first `p` chars.

Popping multiple items from the front can be done via `Vec::drain`.

It might be worth checking whether `Vec::drain` on a zero-sized range is a no-op or not. If it's `O(N)`, then we should have an explicit check at the beginning whether anything needs to be done or not.